precompiles/randomness/src/lib.rs

```diff
@@ -0,0 +1,439 @@
+		let relay_epoch_index =
+			<Runtime as pallet_randomness::Config>::BabeDataGetter::get_epoch_index();
+		Ok(succeed(
+			EvmDataWriter::new().write(relay_epoch_index).build(),
+		))
+	}
+	fn get_request_status(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+		let request_id = handle.read_input()?.read::<U256>()?.low_u64();
+		// record cost of 2 DB reads
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost() * 2)?;
+		let status =
+			if let Some(RequestState { request, .. }) = Pallet::<Runtime>::requests(request_id) {
+				if request.is_expired() {
+					RequestStatus::Expired
+				} else if request.can_be_fulfilled() {
+					RequestStatus::Ready
+				} else {
+					RequestStatus::Pending
+				}
+			} else {
+				RequestStatus::DoesNotExist
+			};
+		Ok(succeed(EvmDataWriter::new().write(status).build()))
+	}
+	fn get_request(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+		let request_id = handle.read_input()?.read::<U256>()?.low_u64();
+		// record cost of 2 DB reads
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost() * 2)?;
+		let RequestState { request, .. } =
+			Pallet::<Runtime>::requests(request_id).ok_or(error("Request Does Not Exist"))?;
+		let status = if request.is_expired() {
+			RequestStatus::Expired
+		} else if request.can_be_fulfilled() {
+			RequestStatus::Ready
+		} else {
+			RequestStatus::Pending
+		};
+		let (
+			randomness_source,
+			fulfillment_block,
+			fulfillment_epoch,
+			expiration_block,
+			expiration_epoch,
+			request_status,
+		) = match request.info {
+			RequestInfo::BabeEpoch(epoch_due, epoch_expired) => (
+				RandomnessSource::RelayBabeEpoch,
+				0u64,
+				epoch_due,
+				0u64,
+				epoch_expired,
+				status,
+			),
+			RequestInfo::Local(block_due, block_expired) => (
+				RandomnessSource::LocalVRF,
+				block_due
+					.try_into()
+					.map_err(|_| revert("block number overflowed u32"))?,
+				0u64,
+				block_expired
+					.try_into()
+					.map_err(|_| revert("block number overflowed u32"))?,
+				0u64,
+				status,
+			),
+		};
+		let (refund_address, contract_address, fee): (Address, Address, U256) = (
+			request.refund_address.into(),
+			request.contract_address.into(),
+			request.fee.into(),
+		);
+		let writer = EvmDataWriter::new()
+			.write(request_id)
+			.write(refund_address)
+			.write(contract_address)
+			.write(fee)
+			.write(request.gas_limit)
+			.write(request.salt)
+			.write(request.num_words)
+			.write(randomness_source)
+			.write(fulfillment_block)
+			.write(fulfillment_epoch)
+			.write(expiration_block)
+			.write(expiration_epoch)
+			.write(request_status)
+			.build();
+		Ok(succeed(writer))
+	}
+	/// Make request for babe randomness one epoch ago
+	fn request_babe_randomness(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+		let mut input = handle.read_input()?;
+		let contract_address = handle.context().caller;
```

Comment:
Where do we ensure this is actually a contract and not an EOA?

Reply:
We do not, it is a TODO in the PR description. How do we ensure this?

Reply:
We may have discussed this already -- we could check for any contract code in storage at the specified account...

Reply:
Note that a contract could self-destruct, though, so maybe that's not a great idea.

Reply:
Agreed, we should make sure fulfilling when an EOA is in the request behaves well, but I think it would be good to prevent this when requesting if at all possible.